File: platforms/desktop/frontend/src/components/settings/sync/__tests__/syncPairing.test.ts
import { describe, expect, it } from 'vitest';
import {
  decodeSyncPairingToken,
  encodeS3SyncPairingToken,
  encodeSyncPairingToken,
  isS3PairingPayload,
  isWebDavPairingPayload,
} from '../syncPairing';

describe('syncPairing', () => {
  const sampleProvider = {
    serverUrl: 'https://dav.jianguoyun.com/dav/',
    remoteRoot: 'Sona',
    username: 'user@example.com',
    password: 'app-password-123',
  };
  const sampleVaultId = 'vault-uuid-456';

  const sampleS3Provider = {
    endpoint: 'https://r2.cloudflarestorage.com',
    region: 'auto',
    bucket: 'my-bucket',
    remoteRoot: 'sona',
    accessKeyId: 'ak-123',
    secretAccessKey: 'sk-456',
    forcePathStyle: false,
  };

  it('encodes and decodes WebDAV pairing tokens successfully', () => {
    const token = encodeSyncPairingToken(sampleProvider, sampleVaultId, true);
    expect(token).toMatch(/^sonasync:\/\/v1\?data=/);

    const decoded = decodeSyncPairingToken(token);
    expect(decoded).not.toBeNull();
    if (decoded && isWebDavPairingPayload(decoded)) {
      expect(decoded.providerId).toBe('webdav');
      expect(decoded.serverUrl).toBe('https://dav.jianguoyun.com/dav/');
      expect(decoded.remoteRoot).toBe('Sona');
      expect(decoded.username).toBe('user@example.com');
      expect(decoded.vaultId).toBe('vault-uuid-456');
      expect(decoded.providerPassword).toBe('app-password-123');
    } else {
      expect.unreachable('expected WebDAV payload');
    }
  });

  it('encodes without provider password when requested', () => {
    const token = encodeSyncPairingToken(sampleProvider, sampleVaultId, false);
    const decoded = decodeSyncPairingToken(token);
    expect(decoded).not.toBeNull();
    if (decoded && isWebDavPairingPayload(decoded)) {
      expect(decoded.providerPassword).toBe('');
      expect(decoded.vaultId).toBe('vault-uuid-456');
    }
  });

  it('encodes and decodes S3 pairing tokens successfully', () => {
    const token = encodeS3SyncPairingToken(sampleS3Provider, sampleVaultId, true);
    expect(token).toMatch(/^sonasync:\/\/v2\?data=/);

    const decoded = decodeSyncPairingToken(token);
    expect(decoded).not.toBeNull();
    if (decoded && isS3PairingPayload(decoded)) {
      expect(decoded.providerId).toBe('s3');
      expect(decoded.endpoint).toBe('https://r2.cloudflarestorage.com');
      expect(decoded.region).toBe('auto');
      expect(decoded.bucket).toBe('my-bucket');
      expect(decoded.remoteRoot).toBe('sona');
      expect(decoded.accessKeyId).toBe('ak-123');
      expect(decoded.secretAccessKey).toBe('sk-456');
      expect(decoded.vaultId).toBe('vault-uuid-456');
      expect(decoded.forcePathStyle).toBe(false);
    } else {
      expect.unreachable('expected S3 payload');
    }
  });

  it('encodes S3 pairing token without secret key when requested', () => {
    const token = encodeS3SyncPairingToken(sampleS3Provider, sampleVaultId, false);
    const decoded = decodeSyncPairingToken(token);
    expect(decoded).not.toBeNull();
    if (decoded && isS3PairingPayload(decoded)) {
      expect(decoded.secretAccessKey).toBe('');
      expect(decoded.vaultId).toBe('vault-uuid-456');
    }
  });

  it('returns null on invalid token strings', () => {
    expect(decodeSyncPairingToken('')).toBeNull();
    expect(decodeSyncPairingToken('invalid-token')).toBeNull();
    expect(decodeSyncPairingToken('sonasync://v1?data=invalid_base64_???')).toBeNull();
    expect(decodeSyncPairingToken(`sonasync://v1?data=${btoa('{"invalid": true}')}`)).toBeNull();
  });
});

File: platforms/android/client/application/src/test/kotlin/com/sona/android/application/sync/SyncPairingTest.kt
package com.sona.android.application.sync

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test

class SyncPairingTest {
    @Test
    fun `encodes and decodes valid pairing token`() {
        val token = encodeSyncPairingToken(
            serverUrl = "https://dav.jianguoyun.com/dav/",
            remoteRoot = "Sona",
            username = "user@example.com",
            vaultId = "vault-42",
            providerPassword = "app-password-123",
        )

        assert(token.startsWith("sonasync://v1?data="))

        val payload = decodeSyncPairingToken(token)
        assertNotNull(payload)
        assertEquals(1, payload?.v)
        assertEquals("https://dav.jianguoyun.com/dav/", payload?.serverUrl)
        assertEquals("Sona", payload?.remoteRoot)
        assertEquals("user@example.com", payload?.username)
        assertEquals("vault-42", payload?.vaultId)
        assertEquals("app-password-123", payload?.providerPassword)
    }

    @Test
    fun `encodes and decodes token without provider password`() {
        val token = encodeSyncPairingToken(
            serverUrl = "https://cloud.example.com/remote.php/dav/files/user/",
            remoteRoot = "SonaVault",
            username = "user",
            vaultId = "vault-99",
        )

        val payload = decodeSyncPairingToken(token)
        assertNotNull(payload)
        assertEquals(1, payload?.v)
        assertEquals("https://cloud.example.com/remote.php/dav/files/user/", payload?.serverUrl)
        assertEquals("SonaVault", payload?.remoteRoot)
        assertEquals("user", payload?.username)
        assertEquals("vault-99", payload?.vaultId)
        assertNull(payload?.providerPassword)
    }

    @Test
    fun `decodes raw base64 string directly`() {
        val json = """{"v":1,"serverUrl":"https://example.com","remoteRoot":"Root","username":"bob","vaultId":"v1"}"""
        val rawBase64 = java.util.Base64.getEncoder().encodeToString(json.toByteArray())

        val payload = decodeSyncPairingToken(rawBase64)
        assertNotNull(payload)
        assertEquals("https://example.com", payload?.serverUrl)
        assertEquals("Root", payload?.remoteRoot)
        assertEquals("bob", payload?.username)
        assertEquals("v1", payload?.vaultId)
    }

    @Test
    fun `rejects invalid or corrupted pairing tokens`() {
        assertNull(decodeSyncPairingToken("invalid-token"))
        assertNull(decodeSyncPairingToken("sonasync://v1"))
        assertNull(decodeSyncPairingToken("sonasync://v1?data="))
        assertNull(decodeSyncPairingToken("sonasync://v1?data=bm90LWpzb24=")) // "not-json"
        assertNull(decodeSyncPairingToken(""))
    }

    @Test
    fun `detects provider preset from server URL`() {
        assertEquals(WellKnownSyncProviderId.NUTSTORE, detectProviderPresetId("https://dav.jianguoyun.com/dav/"))
        assertEquals(WellKnownSyncProviderId.NEXTCLOUD, detectProviderPresetId("https://cloud.example.com/remote.php/dav/files/user/"))
        assertEquals(WellKnownSyncProviderId.INFINICLOUD, detectProviderPresetId("https://teracloud.jp/dav/"))
        assertEquals(WellKnownSyncProviderId.SYNOLOGY, detectProviderPresetId("https://nas.example.com:5006/home/"))
        assertEquals(WellKnownSyncProviderId.ALIST, detectProviderPresetId("https://alist.example.com/dav/"))
        assertEquals(WellKnownSyncProviderId.CUSTOM, detectProviderPresetId("https://webdav.myhost.com/files/"))
    }
}
